pkg/metadata/storage/registry.go

```diff
@@ -168,78 +192,63 @@ func (b *registryBackend) CheckConfig(storage v1alpha1.StorageConfig) error {
 
 // pushImage will push a v1.Image with provided contents
 func (b *registryBackend) pushImage(ctx context.Context, data []byte, fpath string) error {
-	options := b.getOpts(ctx)
+	options, err := b.getOpts(ctx)
+	if err != nil {
+		return err
+	}
 	contents := map[string][]byte{
 		fpath: data,
 	}
 	i, _ := crane.Image(contents)
 	return crane.Push(i, b.src.Ref.Exact(), options...)
 }
 
-func (b *registryBackend) createRegistry() (*containerdregistry.Registry, error) {
-	cacheDir, err := os.MkdirTemp("", "imageset-catalog-registry-")
-	if err != nil {
-		return nil, err
-	}
-
-	logger := logrus.New()
-	logger.SetOutput(ioutil.Discard)
-	nullLogger := logrus.NewEntry(logger)
-
-	return containerdregistry.NewRegistry(
-		containerdregistry.WithCacheDir(cacheDir),
-		containerdregistry.SkipTLS(b.insecure),
-		// The containerd registry impl is somewhat verbose, even on the happy path,
-		// so discard all logger logs}. Any important failures will be returned from
-		// registry methods and eventually logged as fatal errors.
-		containerdregistry.WithLog(nullLogger),
-	)
-}
-
-func (b *registryBackend) unpack(ctx context.Context, path string) error {
-	reg, err := b.createRegistry()
+// exists checks if the image exists
+func (b *registryBackend) exists(ctx context.Context) error {
+	opts, err := b.getOpts(ctx)
 	if err != nil {
-		return fmt.Errorf("error creating container registry: %v", err)
-	}
-	defer reg.Destroy()
-	ref := image.SimpleReference(b.src.Ref.String())
-	if err := reg.Pull(ctx, ref); err != nil {
 		return err
 	}
-	_, err = reg.Labels(ctx, ref)
-	if err != nil {
+	_, err = crane.Manifest(b.src.Ref.Exact(), opts...)
+	if err == nil {
+		return nil
+	}
+	terr, ok := err.(*transport.Error)
+	if !ok {
 		return err
 	}
-	return reg.Unpack(ctx, ref, path)
-}
-
-// exists checks if the image exists
-func (b *registryBackend) exists(ctx context.Context) error {
-	_, err := crane.Manifest(b.src.Ref.Exact(), b.getOpts(ctx)...)
-	var terr *transport.Error
-	switch {
-	case err == nil:
-		return nil
-	case err != nil && errors.As(err, &terr):
```

Comment:
The old composition of `exists()` is less verbose, and you could add a status code check to get the desired behavior
```go
	var terr *transport.Error
	switch {
	case err != nil && errors.As(err, &terr) && terr.StatusCode == 404:
		return ErrMetadataNotExist
	default:
		return err
	}
```